elastic/beats
# Enable keystore for autodiscover static configuration

Close https://github.com/elastic/beats/issues/12597.

## What this PR does
This PR makes use of keystore for containers/pods that are autodiscovered with static configurations only. Hint based configuration should not have access to the keystore for security reasons.

## How to test it manually

1. Create the keystore to store password for REDIS:
```console
./metricbeat keystore create                                                                                 
Created metricbeat keystore
./metricbeat keystore add REDIS_PASSWORD                                                                      
Enter value for REDIS_PASSWORD: 
Successfully updated the keystore
```
1. Enable `autodiscover` with static template:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            contains:
              docker.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: ["${data.host}:6379"]
              password: ${REDIS_PASSWORD}
```
2. Start metricbeat
3. Start a password redis server in a container to be autodiscovred:
```
docker run -p 6379:6379 \
--entrypoint "redis-server" redis --appendonly yes --requirepass 'passpass'
```
Check that REDIS metrics are successfully collected.
Try to start REDIS with a different password so as to make Metricbeat fail.


Now check that hints based autodiscovered containers have not access to the keystore:
1. Configure hint's based autodiscover:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      hints.enabled: true
```
2. Check that hints have no access to the keystore:
```
docker run -p 6379:6379 \
-l co.elastic.metrics/module=redis \
-l co.elastic.metrics/metricsets=info \
-l co.elastic.metrics/password='${REDIS_PASSWORD}' \
-l co.elastic.metrics/hosts='${data.host}:6379' \
--entrypoint "redis-server" redis --appendonly yes --requirepass 'passpass'
```
You should see Metribeat failing to access REDIS since the password is not retrievable.


cc: @exekias
